### PR TITLE
Fix sending of oh_hais on bad sessions

### PR DESCRIPTION
**Pull request by richvdh**
Fix a bunch of bugs in the code which tried to send an oh_hai message when we
got a message with an unknown megolm session.

### CODE REVIEW

File: lib/crypto/index.js [hunk 1]
```diff
@@ -872,13 +873,21 @@ Crypto.prototype.decryptEvent = function(event) {
...
-        if (event.getRoomId !== undefined && event.getSender !== undefined) {
+        // We've got a message for a session we don't have.  Maybe the sender
+        // forgot to tell us about the session.  Remind the sender that we
+        // exist so that they might tell us about the session on their next
+        // send.
+        //
+        // (Alternatively, it might be that we are just looking at
+        // scrollback... at least we rate-limit the m.new_device events :/)
+        //
+        // XXX: this is a band-aid which masks symptoms of other bugs. It would
```
Comment (ara4n) [inline lib/crypto/index.js]: Is the bandaid not useful in papering over scenarios where we upgrade the protocol (e.g. the session id change from 1.2 to 1.3)?

File: lib/crypto/index.js [hunk 2]
```diff
@@ -897,34 +906,40 @@ Crypto.prototype.decryptEvent = function(event) {
  * @param {string} roomId   The ID of the room we want to remind them about.
  */
 Crypto.prototype._sendPingToDevice = function(userId, deviceId, roomId) {
-    if (deviceId === undefined) {
-        deviceId = "*";
```
Comment (ara4n) [inline lib/crypto/index.js]: why have we lost this? it feels like it's needed if deviceId is optional.

File: lib/crypto/index.js (hunk 1 above)
Comment (richvdh) [inline lib/crypto/index.js]: uh, mayyybee? I'm afraid I'm not *entirely* clear on the difference between 1.2 and 1.3; and in any case I don't really think this is a good way to solve that problem.

File: lib/crypto/index.js (hunk 2 above)
Comment (richvdh) [inline lib/crypto/index.js]: deviceId is always included now, so it just felt like cruft to me. I suppose it might as well stay in.

**Pull request merged**